support-frontend/assets/pages/new-contributions-landing/components/PaymentMethodSelector.jsx

```diff
@@ -75,14 +104,62 @@ function PaymentMethodSelector(props: PropTypes) {
+  const noPaymentMethodsErrorMessage =
+    (<GeneralErrorMessage
+      classModifiers={['no-valid-payments']}
+      errorHeading="Payment methods are unavailable"
+      errorReason="all_payment_methods_unavailable"
+    />);
+
+  // having to do this nasty cast because Flow sucks and type guards don't work through .filter
+  const fullExistingPaymentMethods: FullDetailExistingPaymentMethod[] =
+    ((props.existingPaymentMethods || []).filter(isFullDetailExistingPaymentMethod): any);
 
   return (
     <fieldset className={classNameWithModifiers('form__radio-group', ['buttons', 'contribution-pay'])}>
       <legend className="form__legend">Payment method</legend>
 
       { paymentMethods.length ?
         <ul className="form__radio-group-list">
+          {
+            props.contributionType !== 'ONE_OFF' && !props.existingPaymentMethods && (
+              <div className="awaiting-existing-payment-options">
+                <AnimatedDots appearance="medium" />
+              </div>
+            )
+          }
+          {props.contributionType !== 'ONE_OFF' &&
+          fullExistingPaymentMethods.map((existingPaymentMethod: FullDetailExistingPaymentMethod) => (
+            <li className="form__radio-group-item">
+              <input
+                id={`paymentMethodSelector-existing${existingPaymentMethod.billingAccountId}`}
+                className="form__radio-group-input"
+                name="paymentMethodSelector"
+                type="radio"
+                value={existingPaymentMethod}
+                onChange={() => {
+                    props.updatePaymentMethod(mapExistingPaymentMethodToPaymentMethod(existingPaymentMethod));
+                    props.updateSelectedExistingPaymentMethod(existingPaymentMethod);
+                  }}
+                checked={
+                    props.paymentMethod === mapExistingPaymentMethodToPaymentMethod(existingPaymentMethod) &&
+                    props.existingPaymentMethod === existingPaymentMethod
+                  }
+              />
+              <label
+                htmlFor={`paymentMethodSelector-existing${existingPaymentMethod.billingAccountId}`}
+                className="form__radio-group-label has-existing-payment-option-explainer"
+              >
+                <span className="radio-ui" />
+                <span className="radio-ui__label">{getExistingPaymentMethodLabel(existingPaymentMethod)}</span>
+                {getPaymentMethodLogo(mapExistingPaymentMethodToPaymentMethod(existingPaymentMethod))}
+              </label>
+              <div className="existing-payment-option-explainer">
+                  Used for your{' '}
+                {subscriptionsToExplainerList(existingPaymentMethod.subscriptions.map(subscriptionToExplainerPart))}
+              </div>
+            </li>
+            ))}
```

Comment:
this entire bit is pretty nested and hard to parse, i'm  wonderin if it would be worth it to extract into its own component?